nyt-top-20-science-news/README.md

```diff
@@ -0,0 +1,83 @@
+## Requirements:
+This solution aims to fulfill the following requirements, in order of priority:
+- Display a list of articles.
+- Implement a search feature to dynamically filter articles.
+- Search terms should match against the following article attributes:
+    - title
+    - byline
+    - section
+
+## Getting started
+Before you start, please make sure to set up the correct environment variable. At the root of folder (nyt-top-20-sience-news), please create a `.env` file with the following content:
+
+```
+REACT_APP_NYT_API_KEY=<YOUR API KEY HERE>
+```
+You can find an API key in the parent README.md.
+
+Visit the website [here](https://kevb10.github.io/code-challenge/index.html).
+
+- `yarn install`: Installs dependencies for the project.
+- `yarn start`: Runs app in development mode.
+- `yarn test`: Runs the test watcher in an interactive mode.
+- `yarn run deploy`: Builds and deploys app to github pages.
+- `yarn lint`: Run linter. Use `--fix` to fix all fixable issues.
+
+
+## Process
+Here is the initial todo in order:
+- [x] create react app
+- [x] fetch endpoint
+- [x] display basic stuff
+- [x] search/filter fetched list
+- [x] sort by latest update
+- [x] enable linter
+	- can be painful if enabled from the beginning 
+- [x] style
+	-	whip up a very simple design in figma so that I have a visual of what I want to build
+	-	find a free template that matches it
+- [x] update documentation
+- [x] test (very little)
+- [ ] some kind of pagination
+- [ ] [https://developer.nytimes.com/docs/articlesearch-product/1/overview](https://developer.nytimes.com/docs/articlesearch-product/1/overview) you can use their article search api but i’m assuming we only look through top stories
+- [ ] i wasn’t able to find how many API requests we’re allowed to have.
+- [ ] hit the api for updates and only update app when there's a difference found
+
+## Limitations
+- Too many requests:
+
+I actually got this error at some point. Ideally, we would want some kind of cache to prevent the website from making an API request for every page (re)load.  
+- Ambiguity in number of top stories:
+
+Not sure how many stories there are in _top stories_. There were 23 in this cases but is it possible for it to be 100 or 1000? 80% of the time it's _probably_ less than 50. There are several options here:
+	- Infinite scrolling (popular for news websites)
+	- Lazy-loading 
+	- Pagination
+	
+A [good article about these options](https://www.smashingmagazine.com/2016/03/pagination-infinite-scrolling-load-more-buttons/), specifically for ecommerce.
+Ideally I would do infinite scrolling because, besides the performance benefits, it also retains people to keep browsing. With this approach, we know that there _should_ at least be enough articles to cover the full page. Which means, we won't have to load them all at once. So, whether it's 10 or 100000 won't matter since we're loading small chunks at a time.
```

Comment:
we do this for PLP (product listing pages) on shopthrilling.com and when we enabled it we did see a healthy lift in key metrics around time on site and add to cart, which would confirm your assumption about which approach would be best for ecom. however, we also have some interesting edge cases that it creates around redirecting back and how to handle someone being in the middle "page" of a massive list of items. And then there are the perf implications for mobile of adding too many elements to the DOM. Fun little problems to solve.

Reply:
> we do this for PLP (product listing pages) on shopthrilling.com and when we enabled it we did see a healthy lift in key metrics around time on site and add to cart, which would confirm your assumption about which approach would be best for ecom. 

I noticed. It's nice that you had some kind of metric in place to verify this. I'm curious to know if most people access the site on desktop vs mobile. And whether there's a UX that's better for one but isn't for the other.

> however, we also have some interesting edge cases that it creates around redirecting back and how to handle someone being in the middle "page" of a massive list of items. And then there are the perf implications for mobile of adding too many elements to the DOM. Fun little problems to solve.

 Ah, yeah I can see that.

Reply:
One thing I noticed once when I was browsing is that I wanted to get to the footer (to read more about the company and stuff) from a PLP but I couldn't. I kept scrolling until I realized it's not gonna stop lol. So then I had to scroll all the way up. A minor inconvenience but nothing crazy.

Reply:
> …if most people access the site on desktop vs mobile. And whether there's a UX that's better for one but isn't for the other.
mobile traffic beats out desktop by a large margin. but an interesting wrinkle is that conversion % is much higher on desktop vs mobile, which we read as our mobile experience isn't where it needs to be. Luckily, re: best UX, infinite scroll would seem to be the best solution across devices. For mobile, it's the expected experience, and for desktop it's the ideal one.

> …wanted to get to the footer… but I couldn't
I totally understand this frustrating feeling but I also feel like this is somewhat standard for sites that have infinite scroll. do you know of any sites or design patterns that offer a better way?

Reply:
> do you know of any sites or design patterns that offer a better way?

After some research, this seems like a common problem haha. I think there are multiple ways of handling this out there but what worked for them doesn't necessarily mean it would work for us. 
- Google images uses a mix of infinite scrolling & load more. [Example](https://www.google.com/search?q=thrilling&tbm=isch&ved=2ahUKEwjm9NLyv8_wAhWWHt8KHaqtCB8Q2-cCegQIABAA&oq=thrilling&gs_lcp=CgNpbWcQA1AAWABgi7YCaABwAHgAgAEAiAEAkgEAmAEAqgELZ3dzLXdpei1pbWc&sclient=img&ei=D76hYKaxEZa9_Aaq26L4AQ&bih=946&biw=1920): keep scrolling until you see "Show more results". I think it automatically loads 3 times before you see that button. 
- Facebook ran into this issue too. FB is one of those apps where you just can't get away from infinite scrolling. So they moved it to the bottom left (right for IG) in the unscrollable section. 
- Some people have a floater thing on the bottom right that takes them to the top or to the "About" page. 
- Some have a sticky header. I know [shopthrilling](https://shopthrilling.com/) has it (filters too in PLPs)--just not the menu. We could add a ( ? ) icon or something like that to the top right, right along with the wishlist and cart. This would only look okay on desktop though. The even number looks good on mobile. 

💬  I think to figure out whether or not this is even an "issue" worth solving (for now), we need to [understand user activity/clicks](https://www.hotjar.com/) in a heat map fashion (e.g.). If there seems to be a need to get to the footer or to seek some kind of help (other than chat), then we can start testing different solutions with A/B testing and see what works best for us. 

Bonus: The other thing to consider with infinite scrolling is layout shift. For example, IG/FB handles this by anticipating the size of the content, allocating enough space for when the fetched content is ready. Ours wouldn't be too dynamic in size so all we'd need is to have a placeholder image. 

References:
- [Different ideas](https://ux.stackexchange.com/questions/12234/how-can-the-classic-footer-elements-be-integrated-into-an-infinite-scrolling-lay) (Article is dated but some things are still relevant)
- [Pagination + Layout shift](https://addyosmani.com/blog/infinite-scroll-without-layout-shifts/)